node/src/components/linear_chain_sync/operations.rs

```diff
@@ -128,12 +195,42 @@ where
         {
             Ok(FetchedData::FromStorage { item }) => return Ok(Some(item)),
             Ok(FetchedData::FromPeer { item, .. }) => {
-                // TODO: Validate item, ban peer if necessary
                 // Compute the total weight of the validators
-                let _total_weight: U512 = trusted_validator_weights
-                    .iter()
-                    .map(|(_, weight)| *weight)
-                    .sum();
+                let BlockHeaderWithMetadata {
+                    block_header,
+                    block_signatures,
+                } = *item.clone();
+
+                match validate_finality_signatures(
+                    &block_header,
+                    &trusted_validator_weights,
+                    finality_threshold_fraction,
+                    &block_signatures,
+                ) {
+                    Ok(()) => {}
+                    Err(error) => {
+                        warn!(
+                            "Error validating finality signatures from peer. \
+                             peer: {peer:?}, \
+                             error: {error}",
+                            peer = peer.clone(),
+                            error = error
+                        );
+                        // TODO: ban peer
+                        continue;
+                    }
+                }
+
+                // Store the block header
+                effect_builder
+                    .put_block_header_to_storage(Box::new(block_header.clone()))
+                    .await;
+
+                // Store the finality signatures
+                effect_builder
+                    .put_signatures_to_storage(block_signatures.clone())
+                    .await;
```

Comment:
These are two transactions, right? So it could in theory happen that the header gets stored but the signatures don't.
We just need to keep that in mind so we never assume we necessarily have signatures for stored headers.
But I guess we don't assume that anyway? (E.g. when going backwards from the trusted hash, we don't need signatures at all?)

Reply:
> We just need to keep that in mind so we never assume we necessarily have signatures for stored headers.

Yes.

> But I guess we don't assume that anyway? (E.g. when going backwards from the trusted hash, we don't need signatures at all?)

That's right.  It's safe to fetch a block header by hash, even if the node does not have the finality signatures for it yet.